File: scripts/seed-ubs.ts
/**
 * seed-ubs.ts — Importa todas as UBS do Brasil via DATASUS/CNES
 *
 * Uso:
 *   pnpm exec tsx scripts/seed-ubs.ts
 *
 * O que faz:
 *   1. Cria um parceiro-sistema "DATASUS" (User + Partner) se não existir
 *   2. Baixa o catálogo de municípios do IBGE (para nomes de cidades)
 *   3. Consulta a API do CNES filtrando diretamente por tipo de unidade
 *      (tipos 01 = Posto de Saúde, 02 = Centro de Saúde/UBS, 05 = PSF)
 *      usando o parâmetro ?codigo_tipo_unidade=XX — muito mais eficiente
 *      do que baixar todos os estabelecimentos e filtrar localmente.
 *   4. Pula registros sem coordenadas válidas ou desabilitados
 *   5. Apaga os pontos anteriores do parceiro DATASUS e insere os novos
 */

import { PrismaClient } from "@/generated/prisma/client";
import { PrismaPg } from "@prisma/adapter-pg";

const adapter = new PrismaPg({
  connectionString: process.env.DATABASE_URL!,
  ssl: { rejectUnauthorized: false }, // AWS RDS requer SSL; cert não validado (necessário em produção e dev)
});
const prisma = new PrismaClient({ adapter, log: ["error"] });

// A API limita a 20 registros por página independente do valor de limit
const PAGE_SIZE = 20;

// ─── Mapeamento UF code (IBGE) → sigla ───────────────────────────────────────
const UF_MAP: Record<number, string> = {
  11: "RO", 12: "AC", 13: "AM", 14: "RR", 15: "PA", 16: "AP", 17: "TO",
  21: "MA", 22: "PI", 23: "CE", 24: "RN", 25: "PB", 26: "PE", 27: "AL",
  28: "SE", 29: "BA",
  31: "MG", 32: "ES", 33: "RJ", 35: "SP",
  41: "PR", 42: "SC", 43: "RS",
  50: "MS", 51: "MT", 52: "GO", 53: "DF",
};

// Tipos CNES consultados separadamente via filtro na API
const TIPOS_UBS = [1, 2, 5] as const;

// ─── Helpers ─────────────────────────────────────────────────────────────────

async function fetchJson(url: string): Promise<unknown> {
  const res = await fetch(url, {
    headers: { "Accept": "application/json", "User-Agent": "EcoMed/1.0" },
  });
  if (!res.ok) throw new Error(`HTTP ${res.status} — ${url}`);
  return res.json();
}

/** Retorna mapa codigoMunicipio6Digitos → nomeDoMunicipio */
async function buildMunicipioMap(): Promise<Map<number, string>> {
  console.log("📍 Baixando municípios do IBGE...");
  const data = (await fetchJson(
    "https://servicodados.ibge.gov.br/api/v1/localidades/municipios?orderBy=id",
  )) as Array<{ id: number; nome: string }>;
  const map = new Map<number, string>();
  for (const m of data) {
    // CNES usa 6 dígitos; IBGE usa 7. Os primeiros 6 coincidem.
    map.set(Math.floor(m.id / 10), m.nome);
  }
  console.log(`   ${map.size} municípios carregados.`);
  return map;
}

interface CnesRecord {
  codigo_cnes: number;
  nome_razao_social?: string;
  nome_fantasia?: string;
  codigo_tipo_unidade?: number;
  endereco_estabelecimento?: string;
  numero_estabelecimento?: string;
  bairro_estabelecimento?: string;
  codigo_cep_estabelecimento?: string;
  numero_telefone_estabelecimento?: string;
  endereco_email_estabelecimento?: string;
  latitude_estabelecimento_decimo_grau?: number | null;
  longitude_estabelecimento_decimo_grau?: number | null;
  codigo_uf?: number;
  codigo_municipio?: number;
  codigo_motivo_desabilitacao_estabelecimento?: unknown;
}

/**
 * Pagina a API do CNES filtrando por tipo de unidade.
 * A API aceita ?codigo_tipo_unidade=XX e limita a 20 registros por página.
 */
async function* fetchCnesByTipo(tipo: number): AsyncGenerator<CnesRecord[]> {
  let offset = 0;
  let emptyPages = 0;

  while (true) {
    const url =
      `https://apidadosabertos.saude.gov.br/cnes/estabelecimentos` +
      `?limit=${PAGE_SIZE}&offset=${offset}&codigo_tipo_unidade=${String(tipo).padStart(2, "0")}`;

    let data: { estabelecimentos?: CnesRecord[] } | null = null;

    // 3 tentativas com back-off exponencial
    for (let attempt = 1; attempt <= 3; attempt++) {
      try {
        data = (await fetchJson(url)) as { estabelecimentos?: CnesRecord[] };
        break;
      } catch (err) {
        if (attempt === 3) throw err;
        await sleep(1500 * attempt);
      }
    }

    const items = data?.estabelecimentos ?? [];

    if (items.length === 0) {
      emptyPages++;
      if (emptyPages >= 3) break; // 3 páginas vazias consecutivas = fim
      await sleep(500);
      offset += PAGE_SIZE;
      continue;
    }

    emptyPages = 0;
    yield items;

    if (items.length < PAGE_SIZE) break; // última página

    offset += PAGE_SIZE;
    await sleep(200); // gentil com a API
  }
}

function sleep(ms: number) {
  return new Promise((r) => setTimeout(r, ms));
}

// ─── Parceiro-sistema "DATASUS" ───────────────────────────────────────────────

async function getOrCreateDatasusPartner() {
  const DATASUS_EMAIL = "sistema-datasus@ecomed.eco.br";

  let user = await prisma.user.findUnique({ where: { email: DATASUS_EMAIL } });

  if (!user) {
    user = await prisma.user.create({
      data: {
        email: DATASUS_EMAIL,
        name: "DATASUS — Ministério da Saúde",
        role: "ADMIN",
        emailVerified: new Date(),
      },
    });
    console.log("   Usuário DATASUS criado:", user!.id);
  }

  let partner = await prisma.partner.findFirst({ where: { userId: user!.id } });

  if (!partner) {
    partner = await prisma.partner.create({
      data: {
        userId: user!.id,
        cnpj: "00394544000185", // CNPJ público do Ministério da Saúde
        companyName: "Ministério da Saúde — DATASUS",
        tradeName: "Rede Pública de Saúde (UBS)",
        phone: "136",
      },
    });
    console.log("   Parceiro DATASUS criado:", partner.id);
  }

  return partner;
}

// ─── Converte registro CNES → objeto Point ────────────────────────────────────

function cnesToPoint(
  r: CnesRecord,
  partnerId: string,
  municipios: Map<number, string>,
) {
  if (r.codigo_motivo_desabilitacao_estabelecimento) return null;

  const lat = r.latitude_estabelecimento_decimo_grau ?? 0;
  const lng = r.longitude_estabelecimento_decimo_grau ?? 0;

  if (!lat || !lng || Math.abs(lat) < 0.01 || Math.abs(lng) < 0.01) return null;

  const uf = UF_MAP[r.codigo_uf ?? 0] ?? "BR";
  const cidade = municipios.get(r.codigo_municipio ?? 0) ?? "Não informado";
  const nome = (r.nome_fantasia?.trim() || r.nome_razao_social?.trim() || "UBS").slice(0, 255);

  const logradouro = [r.endereco_estabelecimento?.trim(), r.numero_estabelecimento?.trim()]
    .filter(Boolean)
    .join(", ");
  const endereco = [logradouro, r.bairro_estabelecimento?.trim()]
    .filter(Boolean)
    .join(" — ")
    .slice(0, 255) || "Endereço não informado";

  const cep = (r.codigo_cep_estabelecimento ?? "")
    .replace(/\D/g, "").padStart(8, "0").slice(0, 8);

  return {
    partnerId,
    name: nome,
    address: endereco,
    city: cidade,
    state: uf,
    zipCode: cep || "00000000",
    latitude: lat,
    longitude: lng,
    phone: r.numero_telefone_estabelecimento?.trim().slice(0, 20) || null,
    email: r.endereco_email_estabelecimento?.trim().slice(0, 100) || null,
    status: "APPROVED" as const,
    residueTypes: ["medicamentos", "seringas"],
  };
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  console.log("═══════════════════════════════════════════════════════");
  console.log("  EcoMed — Importação de UBS do DATASUS/CNES");
  console.log("═══════════════════════════════════════════════════════\n");

  // 1. Parceiro
  console.log("1. Configurando parceiro DATASUS...");
  const partner = await getOrCreateDatasusPartner();
  console.log(`   Parceiro: ${partner.id}\n`);

  // 2. Municípios
  const municipios = await buildMunicipioMap();
  console.log();

  // 3. Limpa pontos anteriores
  console.log("3. Removendo pontos anteriores do DATASUS...");
  const { count: removidos } = await prisma.point.deleteMany({
    where: { partnerId: partner.id },
  });
  console.log(`   ${removidos} pontos removidos.\n`);

  // 4. Busca por tipo e insere em lotes
  console.log("4. Importando da API CNES por tipo de unidade...");
  console.log("   (tipos: 01=Posto de Saúde, 02=Centro de Saúde/UBS, 05=PSF)\n");

  let totalInserido = 0;
  let totalIgnorado = 0;

  for (const tipo of TIPOS_UBS) {
    console.log(`\n▶ Tipo ${String(tipo).padStart(2, "0")}...`);
    let pageTipo = 0;
    let inseridoTipo = 0;

    for await (const batch of fetchCnesByTipo(tipo)) {
      pageTipo++;
      const toInsert = batch
        .map((r) => cnesToPoint(r, partner.id, municipios))
        .filter((p): p is NonNullable<typeof p> => p !== null);

      totalIgnorado += batch.length - toInsert.length;

      if (toInsert.length > 0) {
        await prisma.point.createMany({ data: toInsert, skipDuplicates: true });
        totalInserido += toInsert.length;
        inseridoTipo += toInsert.length;
      }

      process.stdout.write(
        `\r   Página ${pageTipo} | Inseridas este tipo: ${inseridoTipo} | Total: ${totalInserido}   `,
      );
    }

    console.log(`\n   Tipo ${String(tipo).padStart(2, "0")} concluído: ${inseridoTipo} UBS inseridas.`);
  }

  console.log(`\n\n═══════════════════════════════════════════════════════`);
  console.log(`✅ Concluído!`);
  console.log(`   UBS inseridas : ${totalInserido}`);
  console.log(`   Ignoradas     : ${totalIgnorado} (sem coord. ou desabilitadas)`);
  console.log(`═══════════════════════════════════════════════════════`);
}

main()
  .catch((err) => {
    console.error("\n❌ Erro:", err);
    process.exit(1);
  })
  .finally(() => prisma.$disconnect());

File: scripts/seed_ubs.py
"""
seed_ubs.py — Importa todas as UBS do Brasil via DATASUS/CNES
Requer: psycopg (v3), httpx
Uso dentro do container ecomed-ia:
    python3 /tmp/seed_ubs.py
"""

import argparse, json, os, sys, time, urllib.parse
import httpx
import psycopg
from psycopg.rows import dict_row

# ─── Config ──────────────────────────────────────────────────────────────────

DB_URL_RAW = os.environ.get("DATABASE_URL", "")
_p         = urllib.parse.urlparse(DB_URL_RAW)
DB_PARAMS  = dict(
    host=_p.hostname,
    port=_p.port or 5432,
    dbname=(_p.path or "/ecomed").lstrip("/"),
    user=urllib.parse.unquote(_p.username or ""),
    password=urllib.parse.unquote(_p.password or ""),
    sslmode="require",
)

DATASUS_EMAIL = "sistema-datasus@ecomed.eco.br"
CNPJ_MS       = "00394544000185"
TIPOS_UBS     = {1, 2, 5}   # 01=Posto de Saúde  02=UBS  05=PSF/Saúde da Família

UF_MAP = {
    11:"RO",12:"AC",13:"AM",14:"RR",15:"PA",16:"AP",17:"TO",
    21:"MA",22:"PI",23:"CE",24:"RN",25:"PB",26:"PE",27:"AL",28:"SE",29:"BA",
    31:"MG",32:"ES",33:"RJ",35:"SP",
    41:"PR",42:"SC",43:"RS",
    50:"MS",51:"MT",52:"GO",53:"DF",
}

# ─── Helpers ─────────────────────────────────────────────────────────────────

def cuid_like(prefix: str, n: int) -> str:
    """Gera um ID único simples suficiente para o seed (não precisa ser cuid real)."""
    import hashlib, time as t
    raw = f"{prefix}-{n}-{t.time_ns()}"
    return "c" + hashlib.md5(raw.encode()).hexdigest()[:23]


def fetch_municipios(client: httpx.Client) -> dict[int, str]:
    print("📍 Baixando municípios do IBGE...", flush=True)
    r = client.get(
        "https://servicodados.ibge.gov.br/api/v1/localidades/municipios?orderBy=id",
        timeout=30,
    )
    r.raise_for_status()
    mapa = {}
    for m in r.json():
        mapa[m["id"] // 10] = m["nome"]   # 7 dígitos → 6 dígitos CNES
    print(f"   {len(mapa)} municípios carregados.", flush=True)
    return mapa


def iter_cnes(client: httpx.Client, limit=20):
    """Gera lotes de estabelecimentos da API CNES com retry e back-off."""
    offset, pages_sem_ubs = 0, 0
    while True:
        url = f"https://apidadosabertos.saude.gov.br/cnes/estabelecimentos?limit={limit}&offset={offset}"
        data = None
        for attempt in range(1, 4):
            try:
                r = client.get(url, timeout=30)
                r.raise_for_status()
                data = r.json()
                break
            except Exception as exc:
                if attempt == 3:
                    raise
                time.sleep(attempt * 1.5)

        items = data.get("estabelecimentos") or []
        if not items:
            break

        yield items

        tem_ubs = any(i.get("codigo_tipo_unidade") in TIPOS_UBS for i in items)
        pages_sem_ubs = 0 if tem_ubs else pages_sem_ubs + 1
        if pages_sem_ubs > 100:
            print("\n   100 páginas sem UBS — interrompendo.", flush=True)
            break
        if len(items) < limit:
            break  # última página

        offset += limit
        time.sleep(0.15)

# ─── DB helpers ──────────────────────────────────────────────────────────────

def get_or_create_partner(conn) -> str:
    """Retorna o partnerId do parceiro DATASUS, criando se necessário."""
    with conn.cursor(row_factory=dict_row) as cur:
        # Usuário
        cur.execute('SELECT id FROM "User" WHERE email = %s', (DATASUS_EMAIL,))
        row = cur.fetchone()
        if row:
            user_id = row["id"]
        else:
            user_id = cuid_like("u", 0)
            cur.execute(
                '''INSERT INTO "User" (id, email, name, role, "emailVerified", "referralCode", "createdAt", "updatedAt")
                   VALUES (%s, %s, %s, %s, NOW(), %s, NOW(), NOW())''',
                (user_id, DATASUS_EMAIL, "DATASUS — Ministério da Saúde", "ADMIN", "ECOMED-DATASUS"),
            )
            print(f"   Usuário DATASUS criado: {user_id}", flush=True)

        # Parceiro
        cur.execute('SELECT id FROM "Partner" WHERE "userId" = %s', (user_id,))
        row = cur.fetchone()
        if row:
            partner_id = row["id"]
        else:
            partner_id = cuid_like("p", 0)
            cur.execute(
                '''INSERT INTO "Partner" (id, "userId", cnpj, "companyName", "tradeName", phone)
                   VALUES (%s, %s, %s, %s, %s, %s)''',
                (partner_id, user_id, CNPJ_MS,
                 "Ministério da Saúde — DATASUS",
                 "Rede Pública de Saúde (UBS)", "136"),
            )
            print(f"   Parceiro DATASUS criado: {partner_id}", flush=True)

    conn.commit()
    return partner_id


def delete_old_points(conn, partner_id: str) -> int:
    with conn.cursor() as cur:
        cur.execute('DELETE FROM "Point" WHERE "partnerId" = %s', (partner_id,))
        count = cur.rowcount
    conn.commit()
    return count


def insert_batch(conn, rows: list[dict]):
    if not rows:
        return
    with conn.cursor() as cur:
        cur.executemany(
            '''INSERT INTO "Point"
               (id, "partnerId", name, address, city, state, "zipCode",
                latitude, longitude, phone, email, status, "residueTypes",
                "createdAt", "updatedAt")
               VALUES
               (%(id)s, %(partnerId)s, %(name)s, %(address)s, %(city)s, %(state)s,
                %(zipCode)s, %(latitude)s, %(longitude)s, %(phone)s, %(email)s,
                %(status)s, %(residueTypes)s, NOW(), NOW())
               ON CONFLICT DO NOTHING''',
            rows,
        )
    conn.commit()

# ─── Main ────────────────────────────────────────────────────────────────────

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dry-run", action="store_true")
    parser.add_argument("--limit", type=int)
    parser.add_argument("--city")
    parser.add_argument("--state")
    parser.add_argument("--min-records", type=int, default=100)
    return parser.parse_args()


def main():
    args = parse_args()
    if not DB_URL_RAW:
        sys.exit("❌ DATABASE_URL não definida")
    if args.limit is not None and args.limit <= 0:
        sys.exit("--limit deve ser maior que zero")
    if args.state and args.state.upper() not in set(UF_MAP.values()):
        sys.exit("--state deve ser uma UF brasileira válida")

    print("═" * 55)
    print("  EcoMed — Importação de UBS do DATASUS/CNES (Python)")
    print("═" * 55 + "\n")

    conn = psycopg.connect(**DB_PARAMS, row_factory=dict_row)

    print("1. Configurando parceiro DATASUS...")
    partner_id = get_or_create_partner(conn)
    print(f"   Parceiro: {partner_id}\n")

    with httpx.Client(
        headers={"Accept": "application/json", "User-Agent": "EcoMed/1.0"},
        follow_redirects=True,
    ) as client:
        municipios = fetch_municipios(client)
        print()

        print("3. Baixando e validando em staging (produção ainda não será alterada)...")
        with conn.cursor() as cur:
            cur.execute(
                'CREATE TEMP TABLE "PointImportStage" '
                '(LIKE "Point" INCLUDING DEFAULTS) ON COMMIT PRESERVE ROWS'
            )

        total_ins = total_ign = page = 0

        for batch in iter_cnes(client):
            page += 1
            rows = []

            for r in batch:
                tipo = r.get("codigo_tipo_unidade")
                if tipo not in TIPOS_UBS:
                    continue
                if r.get("codigo_motivo_desabilitacao_estabelecimento"):
                    total_ign += 1
                    continue

                lat = r.get("latitude_estabelecimento_decimo_grau") or 0
                lng = r.get("longitude_estabelecimento_decimo_grau") or 0
                if not lat or not lng or abs(lat) < 0.01 or abs(lng) < 0.01:
                    total_ign += 1
                    continue

                uf     = UF_MAP.get(r.get("codigo_uf") or 0, "BR")
                cidade = municipios.get(r.get("codigo_municipio") or 0, "Não informado")
                if args.state and uf != args.state.upper():
                    continue
                if args.city and cidade.casefold() != args.city.casefold():
                    continue
                nome   = ((r.get("nome_fantasia") or r.get("nome_razao_social") or "UBS").strip())[:255]

                logr   = ", ".join(filter(None, [
                    (r.get("endereco_estabelecimento") or "").strip(),
                    (r.get("numero_estabelecimento") or "").strip(),
                ]))
                end    = (" — ".join(filter(None, [logr, (r.get("bairro_estabelecimento") or "").strip()])))[:255] or "Endereço não informado"
                cep    = "".join(c for c in str(r.get("codigo_cep_estabelecimento") or "") if c.isdigit()).zfill(8)[:8] or "00000000"
                phone  = ((r.get("numero_telefone_estabelecimento") or "").strip() or None)
                if phone:
                    phone = phone[:20]
                email  = ((r.get("endereco_email_estabelecimento") or "").strip() or None)
                if email:
                    email = email[:100]

                rows.append({
                    "id":           cuid_like("pt", total_ins + len(rows)),
                    "partnerId":    partner_id,
                    "name":         nome,
                    "address":      end,
                    "city":         cidade,
                    "state":        uf,
                    "zipCode":      cep,
                    "latitude":     float(lat),
                    "longitude":    float(lng),
                    "phone":        phone,
                    "email":        email,
                    "status":       "APPROVED",
                    "residueTypes": ["medicamentos", "seringas"],
                })

            if args.limit is not None:
                rows = rows[: max(0, args.limit - total_ins)]
            if rows:
                with conn.cursor() as cur:
                    cur.executemany(
                        '''INSERT INTO "PointImportStage"
                           (id, "partnerId", name, address, city, state, "zipCode",
                            latitude, longitude, phone, email, status, "residueTypes",
                            "createdAt", "updatedAt")
                           VALUES
                           (%(id)s, %(partnerId)s, %(name)s, %(address)s, %(city)s, %(state)s,
                            %(zipCode)s, %(latitude)s, %(longitude)s, %(phone)s, %(email)s,
                            %(status)s, %(residueTypes)s, NOW(), NOW())''',
                        rows,
                    )
            total_ins += len(rows)

            if page % 10 == 0 or rows:
                print(f"\r   Página {page:4d} | Inseridas: {total_ins:6d} | Ignoradas: {total_ign:5d}", end="", flush=True)
            if args.limit is not None and total_ins >= args.limit:
                break

    minimum = min(args.min_records, args.limit) if args.limit else args.min_records
    if total_ins < minimum:
        conn.rollback()
        conn.close()
        sys.exit(f"Carga recusada: {total_ins} registros válidos; mínimo exigido: {minimum}")

    with conn.cursor() as cur:
        cur.execute(
            'SELECT COUNT(*) AS total, COUNT(DISTINCT (name, address, city, state)) AS unicos '
            'FROM "PointImportStage"'
        )
        quality = cur.fetchone()
    report = {
        "source": "DATASUS_CNES",
        "valid": total_ins,
        "rejected": total_ign,
        "unique": quality["unicos"],
        "dryRun": args.dry_run,
        "city": args.city,
        "state": args.state,
    }
    print("\n" + json.dumps(report, ensure_ascii=False))

    if args.dry_run:
        conn.rollback()
        conn.close()
        print("Dry-run concluído; produção não foi alterada.")
        return

    # A troca ocorre em uma única transação. Falha no DELETE ou INSERT restaura
    # automaticamente os pontos anteriores. Somente a fonte DATASUS é afetada.
    with conn.cursor() as cur:
        cur.execute('DELETE FROM "Point" WHERE "partnerId" = %s', (partner_id,))
        removidos = cur.rowcount
        cur.execute(
            '''INSERT INTO "Point"
               (id, "partnerId", name, address, city, state, "zipCode", latitude,
                longitude, phone, email, status, "residueTypes", "createdAt", "updatedAt")
               SELECT id, "partnerId", name, address, city, state, "zipCode", latitude,
                      longitude, phone, email, status, "residueTypes", "createdAt", "updatedAt"
               FROM "PointImportStage"'''
        )
    conn.commit()
    conn.close()

    print(f"\n\n{'═'*55}")
    print("✅ Concluído!")
    print(f"   UBS inseridas : {total_ins}")
    print(f"   UBS substituídas: {removidos}")
    print(f"   Ignoradas     : {total_ign}")
    print("═" * 55)


if __name__ == "__main__":
    main()
